src/serac/coefficients/coefficient_extensions.hpp

```diff
@@ -14,41 +14,151 @@
 
 #include <functional>
 #include <memory>
-
+#include <variant>
+#include <vector>
 #include "mfem.hpp"
 
 #include "serac/numerics/expr_template_ops.hpp"
-
+#include "serac/physics/utilities/finite_element_state.hpp"
 /**
  * @brief Functionality that extends current MFEM capabilities
  *
  */
 namespace serac::mfem_ext {
 
+namespace detail {
+// methods for determining index type
+template <typename T>
+struct index_t {
+  using type = std::size_t;
+};
+
+template <typename T>
+struct index_t<mfem::Array<T>> {
+  using type = int;
+};
+
+// methods for determining the size of a container
+template <typename T>
+auto size(T container)
+{
+  return container.size();
+}
+
+template <typename T>
+std::size_t size(mfem::Array<T> container)
+{
+  return container.Size();
+}
+
+// methods for determining type of coefficient evaluations
+template <typename T, typename = void>
+struct eval_t {
+  using type = T;
+};
+
+template <typename T>
+struct eval_t<T, std::enable_if_t<std::is_base_of_v<mfem::Coefficient, T>>> {
+  using type = double;
+};
+
+template <typename T>
+struct eval_t<T, std::enable_if_t<std::is_base_of_v<mfem::VectorCoefficient, T>>> {
+  using type = mfem::Vector;
+};
```

Comment:
I think `eval_t` is needed for the `Transformed*Coefficient` modification which takes in a std::function and corresponding args instead of several fixed constructors.